src/plugins/dashboard/public/application/utils/create_dashboard_container.tsx

```diff
@@ -0,0 +1,522 @@
+  });
+
+  const lastReloadRequestTime = 0;
+  return {
+    id: savedDashboardId || '',
+    filters: data.query.filterManager.getFilters(),
+    hidePanelTitles: appStateData.options.hidePanelTitles,
+    query: data.query.queryString.getQuery(),
+    timeRange: data.query.timefilter.timefilter.getTime(),
+    refreshConfig: data.query.timefilter.timefilter.getRefreshInterval(),
+    viewMode: appStateData.viewMode,
+    panels: embeddablesMap,
+    isFullScreenMode: appStateData.fullScreenMode,
+    isEmptyState:
+      getShouldShowEditHelp(appStateData, dashboardConfig) ||
+      getShouldShowViewHelp(appStateData, dashboardConfig) ||
+      shouldShowUnauthorizedEmptyState(appStateData, services),
+    useMargins: appStateData.options.useMargins,
+    lastReloadRequestTime,
+    title: appStateData.title,
+    description: appStateData.description,
+    expandedPanelId: appStateData.expandedPanelId,
+    timeRestore: appStateData.timeRestore,
+  };
+};
+
+const getChangesForContainerStateFromAppState = (
+  currentContainer: DashboardContainer,
+  appStateDashboardInput: DashboardContainerInput
+) => {
+  if (!currentContainer || isErrorEmbeddable(currentContainer)) {
+    return appStateDashboardInput;
+  }
+
+  const containerInput = currentContainer.getInput();
+  const differences: Partial<DashboardContainerInput> = {};
+
+  // Filters shouldn't  be compared using regular isEqual
+  if (
+    !opensearchFilters.compareFilters(
+      containerInput.filters,
+      appStateDashboardInput.filters,
+      opensearchFilters.COMPARE_ALL_OPTIONS
+    )
+  ) {
+    differences.filters = appStateDashboardInput.filters;
+  }
+
+  Object.keys(containerInput).forEach((key) => {
+    if (key === 'filters') return;
+    const containerValue = (containerInput as { [key: string]: unknown })[key];
+    const appStateValue = ((appStateDashboardInput as unknown) as {
+      [key: string]: unknown;
+    })[key];
+    if (!isEqual(containerValue, appStateValue)) {
+      (differences as { [key: string]: unknown })[key] = appStateValue;
+    }
+  });
+
+  // cloneDeep hack is needed, as there are multiple place, where container's input mutated,
+  // but values from appStateValue are deeply frozen, as they can't be mutated directly
+  return Object.values(differences).length === 0 ? undefined : cloneDeep(differences);
+};
+
+const handleDashboardContainerChanges = (
+  dashboardContainer: DashboardContainer,
+  appState: DashboardAppStateContainer,
+  dashboardServices: DashboardServices,
+  dashboard: Dashboard
+) => {
+  let dirty = false;
+  if (!appState) {
+    return;
+  }
+  const appStateData = appState.getState();
+  const savedDashboardPanelMap: { [key: string]: SavedDashboardPanel } = {};
+  const { opensearchDashboardsVersion } = dashboardServices;
+  const input = dashboardContainer.getInput();
+  appStateData.panels.forEach((savedDashboardPanel) => {
+    if (input.panels[savedDashboardPanel.panelIndex] !== undefined) {
```

Comment:
Is `input.panels[savedDashboardPanel.panelIndex]` being `null` acceptable?

Reply:
Copied from above

======

I have created a comment here: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3365#issuecomment-1627593824.

This logic was ported over, the dashboards plugin is a pretty complex plugin since it's a container embeddable that is managing its state and its children states so previous attempts to clean up certain logic has caused edge case failures.